offload/liboffload/src/OffloadImpl.cpp

```diff
@@ -34,6 +34,69 @@ struct ol_platform_handle_t_ {
   std::vector<ol_device_handle_t_> Devices;
 };
 
+struct ol_queue_handle_t_ {
+  __tgt_async_info *AsyncInfo;
+  ol_device_handle_t Device;
+  std::atomic_uint32_t RefCount;
+};
+
+struct ol_event_handle_t_ {
+  void *EventInfo;
+  ol_queue_handle_t Queue;
+  ol_device_handle_t Device;
```

Comment:
I'm not sure, but I'm afraid these levels of indirection and duplication might cause a problem at some point. Some of it is also not clear to me. The device associated with a queue cannot change, why do we need a device_handle_t that adds a level of indirection (or is it expected to be flattened later)? For an event, why do we need a Device member that is already in Queue (and I don't think ever used)?

Reply:
I agree that adding a level of indirection here isn't ideal. We should aim to flatten it if possible, or at least keep things to a minimum. Ideally these would just be reference-counting wrappers over the underlying objects. It gets a little complicated if, for example, a user queries which device a queue belongs to - you want to return the `ol_device_handle_t` that they used rather than the underlying GenericDeviceTy, so it helps to keep the handles around.

In this case the `Device` was definitely not needed anyway so I've removed it.